File: solutions/213/main.cpp
/* ************************************************************************
> File Name:     main.cpp
> Author:        Sean
> Project:    LeetCode
> Created Time:  Wed Nov 24 11:07:07 2021
> Description:   
 ************************************************************************/
#include <vector> 
#include <gtest/gtest.h> 

using namespace std; 

class Solution {
private:
    int helper(vector<int>& nums, int start, int end) {
        // check input
        if (start == end) return nums[start];
        // dp table
        vector<int> dp(end - start + 1, 0);
        dp[0] = nums[start];
        dp[1] = max(nums[start], nums[start+1]);
        // note, start, end are inclusive
        for (int i = start+2; i <= end; ++i) {
            dp[i-start] = max(dp[i-start-1], dp[i-start-2] + nums[i]);
        }
        return dp[end - start];
    }
public:
    int rob(vector<int>& nums) {
        // check input
        if (nums.size() == 1) return nums[0];
        if (nums.size() == 2) return max(nums[0], nums[1]);
        int profit1 = helper(nums, 0, nums.size()-2);
        int profit2 = helper(nums, 1, nums.size()-1);
        return max(profit1, profit2);
    }
};

TEST(Test213, SimpleTest) {
    vector<int> nums{
        2,3,2
    };
    Solution s; 
    EXPECT_EQ(s.rob(nums), 3);
}

int main() {
    testing::InitGoogleTest();
    return RUN_ALL_TESTS();
}

File: solutions/213/main.py
from typing import List 

class Solution:
    def rob(self, nums: List[int]) -> int:
        # check input
        if len(nums) == 1:
            return nums[0]
        if len(nums) == 2:
            return max(nums[0], nums[1])
        # dp table 1, consider start, do not consider end
        dp1 = [0] * (len(nums)-1)
        dp1[0] = nums[0]
        dp1[1] = max(nums[0], nums[1])
        for i in range(2, len(nums)-1):
            dp1[i] = max(dp1[i-2] + nums[i], dp1[i-1])
        # dp table 2, consider end, do not consider start
        dp2 = [0] * (len(nums)-1)
        dp2[0] = nums[1]
        dp2[1] = max(nums[1], nums[2])
        for i in range(2, len(nums)-1):
            dp2[i] = max(dp2[i-2] + nums[i+1], dp2[i-1])
        return max(dp1[-1], dp2[-1])

if __name__ == "__main__":
    nums = [2,3,2]
    s = Solution()
    assert s.rob(nums) == 3
